### PR TITLE
BUGFIX:TMA-1744 - Store domain users to cache to avoid calling API multiple times

### PR DESCRIPTION
**Pull request by sangtm**
…

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1131/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1013/ : SUCCESS in 11m 46s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1132/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1014/ : SUCCESS in 15m 26s

File: lib/gooddata/models/domain.rb [hunk 1]
```diff
@@ -220,25 +221,29 @@ def users(domain, id = :all, opts = {})
...
         domain = client.domain(domain)
         if id == :all
-          GoodData.logger.warn("Retrieving all users from domain #{domain.name}")
-          all_users = []
-          page_limit = opts[:page_limit] || 1000
-          offset = opts[:offset] || 0
-          loop do
-            begin
-              tmp = client(opts).get("#{domain.uri}/users", params: { offset: offset, limit: page_limit })
+          if @all_domain_users[domain.name].nil?
```
Comment (danh-ung) [inline lib/gooddata/models/domain.rb:224]: Do not cache it here, it will effect all usecases need domain users. Cache it in user filters only.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1144/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1024/ : FAILURE in 11m 13s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1146/ : SUCCESS in 13s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1026/ : SUCCESS in 11m 24s

File: lib/gooddata/models/user_filters/user_filter_builder.rb [hunk 2]
```diff
@@ -307,7 +309,13 @@ def self.create_user_profile_mapping(filters, project_users, options = {})
             found_list[login] = user.links['self'] if user
           end
         else
-          domain_users = domain.users
+          if @all_domain_users[domain.name].nil?
```
Comment (danh-ung) [inline lib/gooddata/models/user_filters/user_filter_builder.rb:312]: please log splunk when domain.users is called, we should log missing_list.size, domain.name, domain_users.size, use_cache=true/false (@all_domain_users[domain.name] != null)

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: lib/gooddata/models/user_filters/user_filter_builder.rb (hunk 2 above)
Comment (sangtm) [reply lib/gooddata/models/user_filters/user_filter_builder.rb:312]: Added logs. Thank you

File: lib/gooddata/models/user_filters/user_filter_builder.rb [hunk 3]
```diff
@@ -307,7 +309,15 @@ def self.create_user_profile_mapping(filters, project_users, options = {})
...
+            domain_users = domain.users
+            @all_domain_users[domain.name] = domain_users
+            GoodData.logger.info("Get all users for domain #{domain.name}, number_users=#{domain_users.size}, number_missing_users=#{missing_list.size}, use_cache=false")
```
Comment (danh-ung) [inline lib/gooddata/models/user_filters/user_filter_builder.rb:315]: log should have an action, e.g action=lcm_get_domain_users to be searchable

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1151/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1031/ : FAILURE in 7m 25s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1153/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1033/ : FAILURE in 11m 00s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1156/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1036/ : SUCCESS in 8m 48s

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1206/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1079/ : SUCCESS in 18m 44s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1208/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1081/ : SUCCESS in 11m 47s

File: lib/gooddata/models/user_filters/user_filter_builder.rb [hunk 4]
```diff
@@ -299,21 +302,42 @@ def self.create_user_profile_mapping(filters, project_users, options = {})
...
+            @mutex.lock
+            if @all_domain_users[domain.name].nil?
+              begin
+                domain_users = domain.users
+                @all_domain_users[domain.name] = domain_users
+                GoodData.logger.info("action=lcm_get_domain_users domain=#{domain.name} number_users=#{domain_users.size} missing_users=#{missing_list.size} use_cache=false")
+              rescue StandardError => e
+                GoodData.logger.warn("action=lcm_get_domain_users domain=#{domain.name} error=#{e}")
+              ensure
+                @mutex.unlock
```
Comment (danh-ung) [inline lib/gooddata/models/user_filters/user_filter_builder.rb:323]: mutex is not unlocked if if @all_domain_users[domain.name] != nill?

Comment (sangtm) [reply lib/gooddata/models/user_filters/user_filter_builder.rb:323]: Only the first thread which is calling API and cached will do the lock/unlock. Other will wait for unlocking then used cached. So we don't need thread with @all_domain_users[domain.name] != nill? unlock the mutex. Thanks anh

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1212/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1085/ : FAILURE in 11m 59s

**Comment by sangtm** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1213/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1086/ : SUCCESS in 12m 29s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1226/ : SUCCESS in 5s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1097/ : FAILURE in 10m 33s

**Comment by sangtm** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1227/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1098/ : SUCCESS in 16m 50s

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ruby-gate-PR1867

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-promote-pipeline/182/ : SUCCESS in 15m 38s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-sonarqube-scan-zuul-gate/103/ : SUCCESS in 4m 36s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-merge-compliance-check/645/ : SUCCESS in 46s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-gitops-deploy-pipeline/68/ : SUCCESS in 2m 07s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-fossa-scanning-zuul-el7/188/ : FAILURE in 3m 24s (non-voting)

**Pull request merged**